X.A.CycleRecentWS: Cycle workspaces, not windowsets

slotThe opened the pull request:
### Description

Fixes #504.  The patch got a little bit more complicated; see below.

The way that workspace cycling is implemented right now—by generating
new windowsets where the things we want happen to be true and then
replacing the old windowsets—is not safe, as workspaces and layouts may
carry state with them.  That state will get lost in translation when
windowsets are simply replaced.

As an example, a conflict occurs when using `X.L.ShowWName.showWName` in
one's layoutHook.  When cycling through workspaces via, e.g.,
`cycleRecentWS` the flashed workspace tag will not disappear, as the
necessary state to control this isn't present in the new windowset.

Instead, what we want to do is to keep the "current" windowset alive and
actually switch to the requested workspaces.  This mostly works without
much trouble, the only hard part is maintaining the invariant that
previewed workspaces don't count towards the history of recently-viewed
workspaces.  This is done by remembering the tag-order of the original
windowset and then restoring that at the end.

This is a breaking change, insofar as it changes the type signatures of
the exported functions `recentWS`, `cycleWindowSets`, and
`toggleWindowSets` to return a list of `WorkspaceId`s instead of a list
of `WindowSet`s.

Fixes: https://github.com/xmonad/xmonad-contrib/issues/504

I wonder about renaming at least `toggleWindowSets`.  The name is not quite accurate anymore and it's already a breaking change so...

@oogeek can you confirm that this works for you?

### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [x] I've considered how to best test these changes (property, unit,
        manually, ...) and concluded: The pure functions are rather testable and it may be worth actually implementing tests for them.

  ~~- [ ] I updated the `XMonad.Doc.Extending` file (if appropriate)~~

TODO:

  - [ ] I updated the `CHANGES.md` file

slotThe commented on XMonad/Actions/CycleRecentWS.hs:
```diff
@@ -76,56 +80,90 @@ toggleRecentNonEmptyWS :: X ()
+               foc <- gets $ tag . workspace . current . windowset
+               -- If there are any inconsistencies when restoring, give
+               -- up and just return a no-op.
+               windows $ \w -> maybe w (view foc) (restoreOrder origWSet w)
```
Okay I cleaned this up a bit to hopefully make it a little bit more readable.  At least on a single monitor this works now; will check on a mutli-monitor setup once have access to that again.

liskin replied:
I added a quickcheck property `restoreOrder ss (view t ss) === ss` and it's falsifiable:

https://x0.at/kHw

(The reason I added it was because I wanted to simplify the code, as complex code is more likely to contain bugs… Anyway, I'll continue working on it, so no need to try to fix it now.)

liskin commented:
I pushed my attempt to simplify the fix.

Note that I don't hate the `MultiWayIf`, I just reverted the entire setOption change as I don't think it's necessary any more. Feel free to add another commit that just reformats it again.

Also note that I haven't bothered to actually test the final result in a real xmonad instance, I developed only against the property test. Can @oogeek please test that it still fixes their issue?

slotThe commented:
@oogeek friendly ping :)

liskin reviewed (changes requested):
Needs a CHANGES entry but otherwise I'd consider this tested and mergeable.

slotThe commented:
Okay I rebased, updated the `CHANGES.md`, and enabled PatternGuards in b65b83661b37c3f29738985e796f5cca17a1440a to make the CI happy (thanks for enabling `-Werror` btw, this will hopefully prevent the buildup of tech debt this time around :))

Will merge once the CI is done and happy
